server.go

```diff
@@ -480,29 +442,94 @@ func (a *App) reply(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	// locking?
-	_, err = a.redis.LPush(a.ctx, "msgbus.system.reply", body).Result()
+	_, err = a.redis.LPush(a.ctx, "msgbus.system.reply", body)
 	if err != nil {
 		err = errors.Wrap(err, "couldn't push entry to reply queue")
 		w.Write(errorReply(err.Error()))
 	}
 	w.Write(successReply())
 
 }
+func (a *App) Serve(ctx context.Context) error {
+	a.ctx = ctx
+	go a.runServer(ctx)
+	var serverErr error
+	go func() {
+		if serverErr = a.server.ListenAndServe(); serverErr != http.ErrServerClosed {
+			log.Error().Err(serverErr).Msg("server ListenAndServe failed")
+		}
+	}()
+	interval := 2 * time.Second
+	timer := time.NewTimer(interval)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug().Msg("shutting down the server")
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+			err := a.Cancel(ctx)
+			if err != nil {
+				log.Err(err)
+			}
+			if serverErr != nil && serverErr != http.ErrServerClosed {
+				return errors.Wrap(serverErr, "server ListenAndServe failed")
+			} else if err != nil {
+				return errors.Wrap(err, "error while stopping server")
+			}
+			return err
+		case <-timer.C:
+			timer.Reset(interval)
+		}
+	}
+}
+
+func (a *App) Cancel(ctx context.Context) error {
+	return a.server.Shutdown(ctx)
+}
 
+func NewServer(debug bool, substrate string, redisServer string, twin int) *App {
+	router := mux.NewRouter()
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     redisServer,
+		Password: "", // no password set
+		DB:       0,  // use default DB
+	})
+	backend := RedisBackend{client: rdb}
+	a := &App{
+		debug: debug,
+		redis: backend,
+		twin:  twin,
+		ctx:   context.Background(),
```

Comment:
This is absolutely not acceptable. A context should never be stored as a state on the object. (what is the point)? The context is "request time" value it should live only during a certain operation and then dies. and should never span multiple "operations" or requests.

Some pseudo code 
```
server = Server()

server.Server(ctx) // passing a context that can be canceled by a condition

func server.Serve(ctx) {
  for {
     if context.IsAlive {
         child := ctx.TimeoutIn(10 seconds)
         handleRequest(child) // we handle this request in maximum of 10 seconds. 
         // notice how a single request can be controlled to not exceed a ceratin amount of time
         // but still canceled if the parent context is canceled. (for example server shutdown)
     }
   }
}
```

So context must be always passed around from your entry method, and not stored on the state. storing on the state is a bad design.

I noticed that u use this to call redis commands, but this is wrong instead the ctx passed to Server is the one that should be passed around (from a method to a method) until it reaches down to the redis cmd. this way on a server shutdown, the full stack will get probably canceled.

If you still have issues with the concept please lets have a call